### PR TITLE
nixos/acme: Backport account missing fixes

### PR DESCRIPTION
**Pull request by m1cr0man**
###### Motivation for this change

Backport of #101482

It was requested by at least one user, and the module is still causing issues for users so this will probably not be the last backport of acme-related changes.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by nh2 — APPROVED** [review]
I'm not an expert in ACME, but from https://github.com/NixOS/nixpkgs/pull/101482#issuecomment-751525285 I can confirm that applying these backports fixed the problem for me on 20.09.

File: nixos/modules/security/acme.xml [hunk 1]
```diff
@@ -263,4 +263,28 @@ chmod 400 /var/lib/secrets/certs.secret
    ones.
   </para>
  </section>
+ <section xml:id="module-security-acme-fix-jws">
+  <title>Fixing JWS Verification error</title>
+
+  <para>
+   It is possible that your account credentials file may become corrupt and need
+   to be regnerated. In this scenario lego will produce the error <literal>JWS verification error</literal>.
```
Comment (flokli) [inline nixos/modules/security/acme.xml:271]:
```suggestion
   to be regenerated. In this scenario lego will produce the error <literal>JWS verification error</literal>.
```

Comment (flokli) [reply nixos/modules/security/acme.xml]: Fixed in master in f71e439688e.

**Comment by flokli** [comment]
@m1cr0man can you cherry-pick -x f71e439688e4c4b231e0f419f81df6610fdaa231 into here?

**Comment by m1cr0man** [comment]
Ah, will do. I was quick to just commit the suggested changes. Will revert that for commit consistency.

**Comment by Ma27** [comment]
Okay, this broke my ACME **again** with the following error:

```
acme-<redacted>-start[1582]: 2021/01/02 11:55:15 Account <redacted> is not registered. Use 'run' to register a new account.
```

Fix was - again - to recreate `/var/lib/acme` and restart all affected services.

**Comment by nh2** [comment]
> Okay, this broke my ACME **again** with the following error:

Just for the record, that's the same error messgae which this PR fixed for me.

**Comment by andir** [comment]
I can second @Ma27's result.. It just broke for me and when  switching system generations it just hangs there :/

```
A dependency job for acme-finished-<domain1>.target failed. See 'journalctl -xe' for details.
A dependency job for acme-finished-<domain2>.target failed. See 'journalctl -xe' for details.
```

**Comment by nh2** [comment]
@andir Does @Ma27's fix also fix it for you?

@m1cr0man Do you know if it's possible to make it such that we can somehow handle this case automatically? It's unfortunate that what fixes some system breaks it for others (at least once) -- and I imagine that even undoing the backport won't help, as  people will still get the problem in the next NixOS release.

**Comment by m1cr0man** [comment]
> Okay, this broke my ACME **again** with the following error:
> 
> ```
> acme-<redacted>-start[1582]: 2021/01/02 11:55:15 Account <redacted> is not registered. Use 'run' to register a new account.
> ```
> 
> Fix was - again - to recreate `/var/lib/acme` and restart all affected services.

I don't have a good guess as to how this is occurring. It could be an instance of the account creation race conditions I suppose (do you have multiple certs with the same account email?). The code in this PR _explicitly_ checks if your account registration files exist - so unless they got deleted between the conditional and lego running then it should just be working.

Hmm. I wonder if lego is creating files in the accounts folder even if it fails to create an account? This could happen if your very first renewal fails. Could you take a quick look at the accounts directory for the affected cert (if you still have it)?

> @m1cr0man Do you know if it's possible to make it such that we can somehow handle this case automatically? It's unfortunate that what fixes some system breaks it for others (at least once) -- and I imagine that even undoing the backport won't help, as people will still get the problem in the next NixOS release.

I'm hopeful that #106857 will cover these cases, but failing that I don't have a root cause for this issue right now, so I can't fix it automatically Blowing away /var/lib/acme or any cert related files automatically would not be a sensible thing to do.

**Comment by pjones** [comment]
Came here to report yet another ACME bug.  After I finally had it all working upgrading to the latest NIxOS 20.09 broke my certs again!

```
Jan 09 23:12:52 kilgrave systemd[1]: Starting Renew ACME certificate for [redacted]...
Jan 09 23:12:53 kilgrave acme-[redacted]-start[590]: 2021/01/09 23:12:53 Account [redacted] is not registered. Use 'run' to register a new account.
```

Is there anything I can provide from the logs or whatnot that will help?

**Comment by m1cr0man** [comment]
> Came here to report yet another ACME bug. After I finally had it all working upgrading to the latest NIxOS 20.09 broke my certs again!
> 
> ```
> Jan 09 23:12:52 kilgrave systemd[1]: Starting Renew ACME certificate for [redacted]...
> Jan 09 23:12:53 kilgrave acme-[redacted]-start[590]: 2021/01/09 23:12:53 Account [redacted] is not registered. Use 'run' to register a new account.
> ```
> 
> Is there anything I can provide from the logs or whatnot that will help?

Yes, can you please list the contents of `/var/lib/acme/accounts/${accountHash}/` and subdirectories for the failing cert? You can get the path by doing `systemctl cat acme-$cert.service | grep accounts`. Also can you please state if you have multiple certs configured on your system?

**Comment by pjones** [comment]
```
[root@kilgrave:~]# ls -lR /var/lib/acme/.lego/accounts/976a0a17bb6f6b07707a
/var/lib/acme/.lego/accounts/976a0a17bb6f6b07707a:
total 4
drwx------ 3 acme wwwrun 4096 Jan  9 23:12 acme-v02.api.letsencrypt.org

/var/lib/acme/.lego/accounts/976a0a17bb6f6b07707a/acme-v02.api.letsencrypt.org:
total 4
drwx------ 3 acme wwwrun 4096 Jan  9 23:12 domains@redacted

/var/lib/acme/.lego/accounts/976a0a17bb6f6b07707a/acme-v02.api.letsencrypt.org/domains@redacted:
total 8
-rwx------ 1 acme wwwrun  215 Jan  9 23:12 account.json
drwx------ 2 acme wwwrun 4096 Jan  9 23:12 keys

/var/lib/acme/.lego/accounts/976a0a17bb6f6b07707a/acme-v02.api.letsencrypt.org/domains@redacted/keys:
total 4
-rwx------ 1 acme wwwrun 227 Jan  9 23:12 domains@redacted.key
```

This machine has two certificates for two different domains using the same email address.  On this machine I also use DNS to verify with ACME and not HTTP.

**Comment by m1cr0man** [comment]
I half expect this to happen with multiple accounts using the same name. Since the check for accounts is a basic "ls" of the accounts folder, if multiple renews run at once (with the same account config) then one of them is bound to create something in the folder and make the other think the account already exists, even if it doesn't really. You might think a more advanced check for the account.json and keys would fix it but still you run into race conditions which have been recorded plenty of times in other tickets. The real solution is the "accounts" targets created in #106857

If you run the failed renew service again, does it work?

**Comment by pjones** [comment]
@m1cr0man It looks like running the service again fixes the issue.  I walked away from my computer for a bit and when I returned the services were fine.  The normal timer must have kicked them off and resolved the issue.

I'd like to offer whatever help I can provide to get this all nailed down.  For my own sanity I really need updates from NixOS 20.09 to be stable.  I'm happy to write tests, help work on the ACME module, or whatever else needs to be done.

**Comment by m1cr0man** [comment]
Honestly? I don't think there's much to worry about. The test suite is very comprehensive and in #106857 I have added a race condition test, which I'm now certain was the cause of your issue. At this point I fully plan to back port that PR too, so that the likes of yourself can benefit from it. I would appreciate a 👍  on that PR if you can :)

**Comment by pjones** [comment]
@m1cr0man 

I figured that since this ACME issue was annoying but harmless I could deploy to my production servers.  Of course that was a mistake.

On a server with a *single* certificate I'm getting this:

```
Jan 11 16:39:50 moriarty systemd[1]: Starting Renew ACME certificate for redacted.com...
Jan 11 16:39:51 moriarty acme-redacted.com-start[5092]: 2021/01/11 16:39:51 [INFO] [redacted.com] acme: Trying renewal with 727 hours remaining
Jan 11 16:39:51 moriarty acme-redacted.com-start[5092]: 2021/01/11 16:39:51 [INFO] [redacted.com] acme: Obtaining bundled SAN certificate
Jan 11 16:39:51 moriarty acme-redacted.com-start[5092]: 2021/01/11 16:39:51 acme: error: 400 :: POST :: https://acme-v02.api.letsencrypt.org/acme/new-order :: urn:ietf:params:acme:error:malformed :: JWS verification error, url:
Jan 11 16:39:51 moriarty systemd[1]: acme-redacted.com.service: Main process exited, code=exited, status=1/FAILURE
```

On servers that have more than one cert they are failing with the "account is not registered" issue.  Trying to run them one at a time does *not* resolve the error.

What info do you need to help me get this sorted out?

**Comment by pjones** [comment]
@m1cr0man 

On the server with multiple certificates, trying to run a single renewal fails with the "account not registered error".  This is what the account directory looks like for that service:

```
$ sudo ls -lR /var/lib/acme/.lego/accounts/976a0a17bb6f6b07707a
/var/lib/acme/.lego/accounts/976a0a17bb6f6b07707a:
total 4
drwx------ 3 acme wwwrun 4096 Jan 11 15:46 acme-v02.api.letsencrypt.org

/var/lib/acme/.lego/accounts/976a0a17bb6f6b07707a/acme-v02.api.letsencrypt.org:
total 4
drwx------ 3 acme wwwrun 4096 Jan 11 15:46 domains@redacted

/var/lib/acme/.lego/accounts/976a0a17bb6f6b07707a/acme-v02.api.letsencrypt.org/domains@redacted:
total 4
drwx------ 2 acme wwwrun 4096 Jan 11 15:46 keys

/var/lib/acme/.lego/accounts/976a0a17bb6f6b07707a/acme-v02.api.letsencrypt.org/domains@redacted/keys:
total 4
-rw-r----- 1 acme wwwrun 227 Jan 11 15:46 domains@redacted.key
```

**Pull request merged**